From fdb3eb6cadd8ae7600153b898471af252380ea57 Mon Sep 17 00:00:00 2001
From: Gajanan Bhat <gajanan.bhat@intel.com>
Date: Fri, 13 Nov 2015 17:40:10 +0530
Subject: [PATCH 4/4] FOR_UPSTREAM [VPG]: drm/i915: Reject color key request
 for invalid planes

Primary and cursor planes also get registered as drm planes. Reject request for
color key set/get for planes other than sprite.
The issue was exposed during graphics card fuzzing. An I/O control on
/dev/dri/controlD64 crashed the device. This patch fixes the issue.

Change-Id: Id8940b0dffb110f5869d59a85bbe20716650c727
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-10857
Signed-off-by: Gajanan Bhat <gajanan.bhat@intel.com>
Signed-off-by: Uma Shankar <uma.shankar@intel.com>
Reviewed-on: https://android.intel.com:443/443498
---
 drivers/gpu/drm/i915/intel_sprite.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_sprite.c b/drivers/gpu/drm/i915/intel_sprite.c
index f6bf42c..7e96324 100644
--- a/drivers/gpu/drm/i915/intel_sprite.c
+++ b/drivers/gpu/drm/i915/intel_sprite.c
@@ -1930,6 +1930,14 @@ int intel_sprite_set_colorkey(struct drm_device *dev, void *data,
 	}
 
 	plane = obj_to_plane(obj);
+
+	/*
+	 * Primary plane also gets registered as drm_plane.
+	 * Apply Color Key only if its sprite plane.
+	 */
+	if (plane->type != DRM_PLANE_TYPE_OVERLAY)
+		return -EINVAL;
+
 	intel_plane = to_intel_plane(plane);
 	ret = intel_plane->update_colorkey(plane, set);
 
@@ -1959,6 +1967,10 @@ int intel_sprite_get_colorkey(struct drm_device *dev, void *data,
 	}
 
 	plane = obj_to_plane(obj);
+
+	if (plane->type != DRM_PLANE_TYPE_OVERLAY)
+		return -EINVAL;
+
 	intel_plane = to_intel_plane(plane);
 	intel_plane->get_colorkey(plane, get);
 
-- 
1.9.1

